### PR TITLE
rafthttp: refactor streamtype

### PR DESCRIPTION
**Pull request by yichengq**
1. Add String() func to represent streamtype better
2. Add parse func to let convertion between endpoint and streamtype
clean

I do this for log cleanup and backoff work.
There doesn't exist an official way to represent stream type well, so i introduce this one.

### CODE REVIEW

File: rafthttp/http.go [hunk 1]
```diff
@@ -153,17 +153,9 @@ func (h *streamHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
-		t = streamTypeMsgApp
-	case path.Join(RaftStreamPrefix, string(streamTypeMsgApp)):
-		t = streamTypeMsgAppV2
-	case path.Join(RaftStreamPrefix, string(streamTypeMessage)):
-		t = streamTypeMessage
-	default:
-		log.Printf("rafthttp: ignored unexpected streaming request path %s", r.URL.Path)
+	t, err := parseStreamType(path.Dir(r.URL.Path))
+	if err != nil {
+		log.Printf("rafthttp: ignored streaming request path %s (%v)", r.URL.Path, err)
```
Comment (xiang90) [inline rafthttp/http.go]: i do not think i can understand this if i see it in the log.

File: rafthttp/stream.go [hunk 2]
```diff
@@ -52,7 +54,17 @@ var (
 	}
 )
 
-type streamType string
+type streamType int
+
+func parseStreamType(path string) (streamType, error) {
```
Comment (xiang90) [inline rafthttp/stream.go]: i do not quite understand why we want this func. why dont we make it a map?

File: rafthttp/stream.go [hunk 3]
```diff
@@ -68,6 +80,19 @@ func (t streamType) endpoint() string {
...
+func (t streamType) String() string {
+	switch t {
+	case streamTypeMsgApp:
+		return "stream MsgApp"
+	case streamTypeMsgAppV2:
+		return "stream MsgApp v2"
+	case streamTypeMessage:
+		return "stream Message"
+	default:
+		return fmt.Sprintf("stream type %d", int(t))
```
Comment (xiang90) [inline rafthttp/stream.go]: why do we need the default case?

**Comment by xiang90** [comment]
I did not see the benefits of this pull request actually. Can you please explain more?

**Comment by yichengq** [comment]
I want a method to represent a stream type well in log. Today we print it out using `string(t)`, which prints out `msgapp`/`msgappv2`/`message`. This is hard to understand without stream context in log line.

**Comment by xiang90** [comment]
@yichengq I think we can simply add a string func to the existing code.

**Comment by yichengq** [comment]
@xiang90 k. will use that simple way.

**Pull request closed without merging**